README.md
# Stormpath PHP SDK 

[![Build Status](https://api.travis-ci.org/stormpath/stormpath-sdk-php.svg?branch=master,dev)](https://travis-ci.org/stormpath/stormpath-sdk-php) 
[![Codecov](https://img.shields.io/codecov/c/github/stormpath/stormpath-sdk-php.svg)](https://codecov.io/github/stormpath/stormpath-sdk-php)
[![Total Downloads](https://poser.pugx.org/stormpath/sdk/d/total.svg)](https://packagist.org/packages/stormpath/sdk)
[![Latest Stable Version](https://poser.pugx.org/stormpath/sdk/v/stable.svg)](https://packagist.org/packages/stormpath/sdk)
[![License](https://poser.pugx.org/stormpath/sdk/license.svg)](https://packagist.org/packages/stormpath/sdk)
[![Chat](https://img.shields.io/badge/chat-on%20freenode%20-green.svg)](http://webchat.freenode.net/?channels=#stormpath)
[![Support](https://img.shields.io/badge/support-support@stormpath.com-blue.svg)](mailto:support@stormpath.com?subject=Stormpath+PHP+SDK)

[Stormpath](https://stormpath.com) is a complete user management API.  This
library gives your PHP application access to all of Stormpath's features:

- Robust authentication and authorization.
- Schemaless user data and profiles.
- A hosted login subdomain, for easy Single Sign-On across your apps.
- External login with social providers like Facebook and Google, or SAML IdPs.
- Secure API key authentication for your service.


## Installation

**stormpath-sdk-php** is available on Packagist as the [stormpath/sdk](http://packagist.org/packages/stormpath/sdk) package.

Run `composer require stormpath/sdk` from the root of your project in terminal, and you are done.


## Quickstart

To learn how to use the Stompath PHP SDK in a simple project, follow our quickstart:

* [PHP Quickstart](http://docs.stormpath.com/php/product-guide/latest/quickstart.html)

#### Full Documentation

We have moved our full documentation away from the Github readme file.  For full documentation, please visit [our new documentation](http://docs.stormpath.com/php/product-guide/latest)


## Testing

The PHP SDK uses `phpunit` for testing.  These tests are full integration tests which means it hits actual endpoints of the API.

To setup testing, first, clone the repository.  You will not be able to run the tests from your vendor folder as all unnecessary items are removed
when you require the SDK with composer.  You will also need to set an environment variable to set up your API keys. 

#### On Mac
    
    export STORMPATH_SDK_TEST_API_KEY_FILE_LOCATION=path_to_apiKey.properties_file

#### On Windows
  
    setx STORMPATH_SDK_TEST_API_KEY_FILE_LOCATION path_to_apiKey.properties_file
    
After you have this set, you will be able to run `phpunit` from the root of the SDK folder.  If your API keys are not from a subscription that has access to 
the SAML resources, you will have to make sure you skip those tests.  You can do so by running `phpunit --exclude-group=saml` from the command line.

## Contributing

Contributions, bug reports, and issues are very welcome! Stormpath regularly maintains this repository, and are quick to review pull requests and accept changes.

You can make your own contributions by forking the develop branch of this repository, making your changes, and issuing pull requests against the `develop` branch.

#### Continuous Integration (Travis CI)
Please note that due to security reasons, travis will not run pull requests submitted.  With your pull request, please submit the results of the tests in the comments.

#### Documentation
If you feel the contributions require document changes as well, or the contributions you want to make are for documentation, please submit a PR to [our documentation repo](https://github.com/stormpath/stormpath-documentation)


## Copyright

Copyright &copy; 2013-2017 Stormpath, Inc. and contributors.

This project is open-source via the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0).

CHANGES.md
stormpath-sdk-php Changelog
===========================


Version 1.19.0
--------------

Released on January 17, 2017

 - Added ability to set ttl and tti to cache.
 
 Version 1.18.0
 --------------
 
 Released on January 3, 2017
 
  - Happy New Year!
  - Added OrgNameKey to Password Grant Type (Resolves #61)
  - Update all copyright dates to 2017
 
Version 1.17.2
--------------

Released on Dec 20, 2016

 - Update JWT Library to 4.0
 - Change all test emails to testmail.stormpath.com
 
Version 1.17.1
--------------

Released on October 26, 2016

 - Fixed bug where Integration was not always reported correctly
 
 Version 1.17.0
 --------------
 
 Released on October 20, 2016
 
  - Added MFA 
  - Converted HTTP requests to using psr-7
  - General Updates and bug fixes

Version 1.16.0
--------------

Released on July 20, 2016 
 
 - Fix cache key for resources 
 - Adds method for getting access to the password modified at property off account
 - Work with Password Policy from Directory
 - Password Reuse policy
 - Adds ability to work with all email templates
 - Now able to add an account directly from the organization resource
 - Create and work with email blacklist and whitelist.
 - Request ID now part of the resource error and error
 - Name and Description can be added to API Keys

Version 1.15.0
--------------

Released on June 14, 2016 

 - PSR-6 Caching (Resolves #130, Fixes #24, Fixes #48)
 - SAML Tests grouped for easy developer plan testing (Fixes #131)
 - Update Readme for moved docs to http://docs.stormpath.com/php/product-guide/latest/
 
 
Version 1.14.0
--------------

Released on March 16, 2016 

 - Documentation updates
 - Using openssl for apiKey options (Fixes #132)
 - Added Github Templates for PRs and Issues (Fixes #138)
 - Passing code to Exception off of Resource Exception (Fixes #138)
 - Update Facebok Provider to include code and access_token (Fixes #142)
 - Official Support for PHP 7 (Fixes #144)
 - Updated PHP Sec Library (Fixes #144)
 - Added .gitattributes for cleaner composer (Fixes #123)
 
Version 1.13.0
--------------

Released on January 26, 2016

 - Added SAML to the SDK!
 - Add ability to use the password reset token from the normal workflow inside of the ID Site
 - Updated Copyright Dates in all files.
 - Updated travis notification to send message to HipChat channel
 

Version 1.12.2
--------------

Released on December 16, 2015

 - Fixed a bug where the accessToken and refreshTokens were not accessible from the Account
 - We now show the refresh token object from OauthGrantAuthenticatorResult for all requests if enabled
 - Removed logging from the phpunit tests. If code coverage is needed, you should now use the flag on phpunit
 - Update Readme Documentation to reference the fixes and phpunit changes
 - Removed branch alias from composer.json file

Version 1.12.1
--------------

Released on December 15, 2015

 - Social Providers isNewAccount was always returning null.  This has been fixed (Fixes Issue #109)
 - Handle ID Site would throw new exception if account was not valid, This prevented you from accessing state or any other properties
 You can not access all properties, and account will just remain null. (Fixes Issue #110)
 - Documentation issue on API Keys and Account for Request Authenticators. (Fixes Issue #105)
 - Update the Base Test so it is not treated as a test. 

Version 1.12.0
--------------

Released on November 5, 2015

- GA Release of the SDK ... YAY!
- Addition of Organization Resource. Fixes Issue #95
- Ability to specify nameKey during ID Site. Fixes Issue #93
- Give you ability to return sptoken object from Password Reset Email. Fixes Issue #102
- Reset Password during Password Reset Workflow. Fixes Issue #103
- Makes the SDK PSR-4 compliant. Fixes Issue #17


Version 1.11.0.beta
------------------

Released on October 29, 2015

- Add Integration flag to the Client. Set it from Integration with static call.
- Token Management with the application. Handles Issue #59
- Token Authentication. Lets Stormpath generate access_token and refresh_token for you from the oauth/token endpint. Handles Issue #59
- Exchange ID Site JWT with an Access Token. Handles Issue #94
- Add 2 new providers, Linkedin and Github. Handles Issue #90
- Documentation change for building a Client. Handles Issue #78

Version 1.10.0.beta
------------------

Released on September 1, 2015

- Ability to change Authentication Scheme.  Fixes Issue #12
- Better error handling in ID Site callback.  Handles Issue #82


Version 1.9.0.beta
------------------

Released on August 10, 2015

- Password Import is now tested for and officially supported 
- New method added for getting size of collection.
- Tests implemented for getting a resource based on an HTML fragment
- Resending an email verification to a user
- Added support for modifying registration workflows
- Removed PHP 5.3 Support
- Updated code coverage and updated PHPUnit package



Version 1.8.1.beta
------------------

Released on August 6, 2015

- Fixed bug with verifyEmailToken method.  This will now allow the return of an Account after verification


Version 1.8.0.beta
------------------

Released on July 30, 2015

- Added integration for Coveralls.io
- Implemented API Key generation for accounts
- Implemented API Key authorization


Version 1.7.0.beta
------------------

Released on July 29, 2015

- Upgraded to PHP-JWT to 2.2.*
- Changed to guzzle/guzzle from guzzle/html


Version 1.6.0.beta
------------------

Released on July 14, 2015

- Fixed a bug when adding custom data to an existing property (like adding an entry to an existing array).
- Added minor documentation changes.


Version 1.5.0.beta
------------------

Released on June 12, 2015

- Implemented social integration with Google and Facebook.
- Fixed a bug when saving custom data after retrieving it from a resource.


Version 1.4.0.beta
------------------

Released on June 3, 2015


src/Util/UUID.php
<?php
/**
 * Copyright 2017 Stormpath, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Stormpath\Util;

/*-
 * Copyright (c) 2011 Fredrik Lindberg - http://www.shapeshifter.se
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *	  notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *	  notice, this list of conditions and the following disclaimer in the
 *	  documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * Alternative this software might be licensed under the following license
 *
 *  Copyright 2011 Fredrik Lindberg
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

/*
 * UUID (RFC4122) Generator
 * http://tools.ietf.org/html/rfc4122
 *
 * Implements version 1, 3, 4 and 5
 */

/*
 * UUID (RFC4122) Generator
 * http://tools.ietf.org/html/rfc4122
 *
 * Implements version 1, 3, 4 and 5
 */
// @codeCoverageIgnoreStart
class UUID
{
    /* UUID versions */
    const UUID_TIME = 1;    /* Time based UUID */
    const UUID_NAME_MD5 = 3;    /* Name based (MD5) UUID */
    const UUID_RANDOM = 4;    /* Random UUID */
    const UUID_NAME_SHA1 = 5;    /* Name based (SHA1) UUID */

    /* UUID formats */
    const FMT_FIELD = 100;
    const FMT_STRING = 101;
    const FMT_BINARY = 102;
    const FMT_QWORD = 1;    /* Quad-word, 128-bit (not impl.) */
    const FMT_DWORD = 2;    /* Double-word, 64-bit (not impl.) */
    const FMT_WORD = 4;    /* Word, 32-bit (not impl.) */
    const FMT_SHORT = 8;    /* Short (not impl.) */
    const FMT_BYTE = 16;    /* Byte */
    const FMT_DEFAULT = 16;

    /* Field UUID representation */
    private static $m_uuid_field = array(
        'time_low' => 0,        /* 32-bit */
        'time_mid' => 0,        /* 16-bit */
        'time_hi' => 0,            /* 16-bit */
        'clock_seq_hi' => 0,        /*  8-bit */
        'clock_seq_low' => 0,        /*  8-bit */
        'node' => array(),        /* 48-bit */
    );

    private static $m_generate = array(
        self::UUID_TIME => 'generateTime',
        self::UUID_RANDOM => 'generateRandom',
        self::UUID_NAME_MD5 => 'generateNameMD5',
        self::UUID_NAME_SHA1 => 'generateNameSHA1',
    );

    private static $m_convert = array(
        self::FMT_FIELD => array(
            self::FMT_BYTE => 'conv_field2byte',
            self::FMT_STRING => 'conv_field2string',
            self::FMT_BINARY => 'conv_field2binary',
        ),
        self::FMT_BYTE => array(
            self::FMT_FIELD => 'conv_byte2field',
            self::FMT_STRING => 'conv_byte2string',
            self::FMT_BINARY => 'conv_byte2binary',
        ),
        self::FMT_STRING => array(
            self::FMT_BYTE => 'conv_string2byte',
            self::FMT_FIELD => 'conv_string2field',
            self::FMT_BINARY => 'conv_string2binary',
        ),
    );

    /* Swap byte order of a 32-bit number */
    private static function swap32($x)
    {
        return (($x & 0x000000ff) << 24) | (($x & 0x0000ff00) << 8) |
            (($x & 0x00ff0000) >> 8) | (($x & 0xff000000) >> 24);
    }

    /* Swap byte order of a 16-bit number */
    private static function swap16($x)
    {
        return (($x & 0x00ff) << 8) | (($x & 0xff00) >> 8);
    }

    /* Auto-detect UUID format */
    private static function detectFormat($src)
    {
        if (is_string($src)) {
            return self::FMT_STRING;
        } elseif (is_array($src)) {
            $len = count($src);
            if ($len == 1 || ($len % 2) == 0) {
                return $len;
            } else {
                return -1;
            }
        } else {
            return self::FMT_BINARY;
        }
    }

    /*
      * Public API, generate a UUID of 'type' in format 'fmt' for
      * the given namespace 'ns' and node 'node'
      */
    public static function generate($type, $fmt = self::FMT_BYTE,
                                    $node = '', $ns = '')
    {
        $func = self::$m_generate[$type];
        if (!isset($func)) {
            return null;
        }
        $conv = self::$m_convert[self::FMT_FIELD][$fmt];

        $uuid = self::$func($ns, $node);

        return self::$conv($uuid);
    }

    /*
      * Public API, convert a UUID from one format to another
      */
    public static function convert($uuid, $from, $to)
    {
        $conv = self::$m_convert[$from][$to];
        if (!isset($conv)) {
            return $uuid;
        }

        return self::$conv($uuid);
    }

    /*
      * Generate an UUID version 4 (pseudo random)
      */
    private static function generateRandom($ns, $node)
    {
        $uuid = self::$m_uuid_field;

        $uuid['time_hi'] = (4 << 12) | (mt_rand(0, 0x1000));
        $uuid['clock_seq_hi'] = (1 << 7) | mt_rand(0, 128);
        $uuid['time_low'] = mt_rand(0, 0xffff) + (mt_rand(0, 0xffff) << 16);
        $uuid['time_mid'] = mt_rand(0, 0xffff);
        $uuid['clock_seq_low'] = mt_rand(0, 255);
        for ($i = 0; $i < 6; ++$i) {
            $uuid['node'][$i] = mt_rand(0, 255);
        }

        return $uuid;
    }

    /*
      * Generate UUID version 3 and 5 (name based)
      */
    private static function generateName($ns, $node, $hash, $version)
    {
        $ns_fmt = self::detectFormat($ns);
        $field = self::convert($ns, $ns_fmt, self::FMT_FIELD);

        /* Swap byte order to keep it in big endian on all platforms */
        $field['time_low'] = self::swap32($field['time_low']);
        $field['time_mid'] = self::swap16($field['time_mid']);
        $field['time_hi'] = self::swap16($field['time_hi']);

        /* Convert the namespace to binary and concatenate node */
        $raw = self::convert($field, self::FMT_FIELD, self::FMT_BINARY);
        $raw .= $node;

        /* Hash the namespace and node and convert to a byte array */
        $val = $hash($raw, true);
        $tmp = unpack('C16', $val);
        foreach (array_keys($tmp) as $key) {
            $byte[$key - 1] = $tmp[$key];
        }

        /* Convert byte array to a field array */
        $field = self::conv_byte2field($byte);

        $field['time_low'] = self::swap32($field['time_low']);
        $field['time_mid'] = self::swap16($field['time_mid']);
        $field['time_hi'] = self::swap16($field['time_hi']);

        /* Apply version and constants */
        $field['clock_seq_hi'] &= 0x3f;
        $field['clock_seq_hi'] |= (1 << 7);
        $field['time_hi'] &= 0x0fff;
        $field['time_hi'] |= ($version << 12);

        return $field;
    }
    private static function generateNameMD5($ns, $node)
    {
        return self::generateName($ns, $node, 'md5',
            self::UUID_NAME_MD5);
    }
    private static function generateNameSHA1($ns, $node)
    {
        return self::generateName($ns, $node, 'sha1',
            self::UUID_NAME_SHA1);
    }

    /*
      * Generate UUID version 1 (time based)
      */
    private static function generateTime($ns, $node)
    {
        $uuid = self::$m_uuid_field;

        /*
           * Get current time in 100 ns intervals. The magic value
           * is the offset between UNIX epoch and the UUID UTC
           * time base October 15, 1582.
           */
        $tp = gettimeofday();
        $time = ($tp['sec'] * 10000000) + ($tp['usec'] * 10) +
            0x01B21DD213814000;

        $uuid['time_low'] = $time & 0xffffffff;
        /* Work around PHP 32-bit bit-operation limits */
        $high = intval($time / 0xffffffff);
        $uuid['time_mid'] = $high & 0xffff;
        $uuid['time_hi'] = (($high >> 16) & 0xfff) | (self::UUID_TIME << 12);

        /*
           * We don't support saved state information and generate
           * a random clock sequence each time.
           */
        $uuid['clock_seq_hi'] = 0x80 | mt_rand(0, 64);
        $uuid['clock_seq_low'] = mt_rand(0, 255);

        /*
           * Node should be set to the 48-bit IEEE node identifier, but
           * we leave it for the user to supply the node.
           */
        for ($i = 0; $i < 6; ++$i) {
            $uuid['node'][$i] = ord(substr($node, $i, 1));
        }

        return $uuid;
    }

    /* Assumes correct byte order */
    private static function conv_field2byte($src)
    {
        $uuid[0] = ($src['time_low'] & 0xff000000) >> 24;
        $uuid[1] = ($src['time_low'] & 0x00ff0000) >> 16;
        $uuid[2] = ($src['time_low'] & 0x0000ff00) >> 8;
        $uuid[3] = ($src['time_low'] & 0x000000ff);
        $uuid[4] = ($src['time_mid'] & 0xff00) >> 8;
        $uuid[5] = ($src['time_mid'] & 0x00ff);
        $uuid[6] = ($src['time_hi'] & 0xff00) >> 8;
        $uuid[7] = ($src['time_hi'] & 0x00ff);
        $uuid[8] = $src['clock_seq_hi'];
        $uuid[9] = $src['clock_seq_low'];

        for ($i = 0; $i < 6; ++$i) {
            $uuid[10 + $i] = $src['node'][$i];
        }

        return $uuid;
    }

    private static function conv_field2string($src)
    {
        $str = sprintf(
            '%08x-%04x-%04x-%02x%02x-%02x%02x%02x%02x%02x%02x',
            ($src['time_low']), ($src['time_mid']), ($src['time_hi']),
            $src['clock_seq_hi'], $src['clock_seq_low'],
            $src['node'][0], $src['node'][1], $src['node'][2],
            $src['node'][3], $src['node'][4], $src['node'][5]);

        return $str;
    }

    private static function conv_field2binary($src)
    {
        $byte = self::conv_field2byte($src);

        return self::conv_byte2binary($byte);
    }

    private static function conv_byte2field($uuid)
    {
        $field = self::$m_uuid_field;
        $field['time_low'] = ($uuid[0] << 24) | ($uuid[1] << 16) |
            ($uuid[2] << 8) | $uuid[3];
        $field['time_mid'] = ($uuid[4] << 8) | $uuid[5];
        $field['time_hi'] = ($uuid[6] << 8) | $uuid[7];
        $field['clock_seq_hi'] = $uuid[8];
        $field['clock_seq_low'] = $uuid[9];

        for ($i = 0; $i < 6; ++$i) {
            $field['node'][$i] = $uuid[10 + $i];
        }

        return $field;
    }

    public static function conv_byte2string($src)
    {
        $field = self::conv_byte2field($src);

        return self::conv_field2string($field);
    }

    private static function conv_byte2binary($src)
    {
        $raw = pack('C16', $src[0], $src[1], $src[2], $src[3],
            $src[4], $src[5], $src[6], $src[7], $src[8], $src[9],
            $src[10], $src[11], $src[12], $src[13], $src[14], $src[15]);

        return $raw;
    }

    private static function conv_string2field($src)
    {
        $parts = sscanf($src, '%x-%x-%x-%x-%02x%02x%02x%02x%02x%02x');
        $field = self::$m_uuid_field;
        $field['time_low'] = ($parts[0]);
        $field['time_mid'] = ($parts[1]);
        $field['time_hi'] = ($parts[2]);
        $field['clock_seq_hi'] = ($parts[3] & 0xff00) >> 8;
        $field['clock_seq_low'] = $parts[3] & 0x00ff;
        for ($i = 0; $i < 6; ++$i) {
            $field['node'][$i] = $parts[4 + $i];
        }

        return $field;
    }

    private static function conv_string2byte($src)
    {
        $field = self::conv_string2field($src);

        return self::conv_field2byte($field);
    }

    private static function conv_string2binary($src)
    {
        $byte = self::conv_string2byte($src);

        return self::conv_byte2binary($byte);
    }

    public static function v4()
    {
        return sprintf('%04x%04x-%04x-%04x-%04x-%04x%04x%04x',

            // 32 bits for "time_low"
            mt_rand(0, 0xffff), mt_rand(0, 0xffff),

            // 16 bits for "time_mid"
            mt_rand(0, 0xffff),

            // 16 bits for "time_hi_and_version",
            // four most significant bits holds version number 4
            mt_rand(0, 0x0fff) | 0x4000,

            // 16 bits, 8 bits for "clk_seq_hi_res",
            // 8 bits for "clk_seq_low",
            // two most significant bits holds zero and one for variant DCE1.1
            mt_rand(0, 0x3fff) | 0x8000,

            // 48 bits for "node"
            mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
        );
    }
}
// @codeCoverageIgnoreEnd
